packages/channel-provider/src/messaging-service.ts

```diff
@@ -90,21 +90,20 @@ export class MessagingService {
     callback?: (result: ResultType) => void
   ) {
     const listener = (event: MessageEvent) => {
-      if (
-        event.data &&
-        event.data.jsonrpc &&
-        isJsonRpcResponse(event.data) &&
-        event.data.id === message.id
-      ) {
-        if (callback) {
-          callback(event.data.result);
+      if (event.data && event.data.jsonrpc) {
+        if (isJsonRpcResponse(event.data) && event.data.id === message.id) {
+          if (callback) {
+            callback(event.data.result);
+          }
+          this.acknowledge();
+          window.removeEventListener('message', listener);
+          logger.info({response: event.data}, 'Received response');
+          resolve(event.data.result);
+        } else if (isJsonRpcErrorResponse(event.data)) {
+          reject(new Error(`Wallet Error ${event.data.error.code}: ${event.data.error.message}`));
```

Comment:
I think this is incorrect: what if it is a `isJsonRpcErrorResponse`, but for a different message (with a different `id`)?